aten/src/ATen/native/Convolution.cpp

```diff
@@ -531,6 +538,14 @@ at::Tensor _convolution(
 
   check_shape_forward(input, weight, bias, params, input_is_mkldnn);
 
+  // I have a choice here to move this above or under the views below.
+  // Temporarily put it above as this is the most generic version,
+  // could potentially move to the end of function if XLA doesn't handle
+  // it.
+  if (input.device() == at::kXLA) {
+    return at::convolution_generic(input, weight, bias, params.stride, params.padding, params.dilation, params.transposed, params.output_padding, params.groups);
+  }
```

Comment:
Yeah, I think my preference is to not hard code XLA here. Instead, we know that e.g. cudnn only handles cuda, and thnn only handles cpu/cuda; everything else will naturally fall through in the end. This is OK for a PoC to show this works with XLA, but I'd prefer we fix it correctly before merging (as this patch will set precedent for future `generic_foo` patches)

Reply:
I somehow feel that this "de-virtualizing" things landed us to "virtual by hand" (if-instance-type-do-this) implementations.